README.md
# koa-happy-router

`koa-happy-router` provides easy-to-use routing configuraion likes `hapi`, powerful and freedom middleware mechanism.

## 🆕 Features

- Easy-to-use routing configuraion
- Powerful and freedom middleware mechanism
- Stronge error handler

## 📦 Install

If you use npm

```bash
npm install koa-happy-router
```

or yarn

```bash
yarn add koa-happy-router
```

## 🎮 Usage

### Basic Usage

```javascript
import Koa from "koa";
import HappyRouter from "koa-happy-router";

const app = new Koa();
const router = new HappyRouter();

router.addRoutes([
  {
    url: "/",
    handler: (ctx) => (ctx.body = "Hello World!"),
  },
]);

app.use(router.routes()).use(router.allowMethods());
app.listen(8080);
```

### Middlewares Usage

You can register the global middleware by `registerMiddlewares`, which can be used in routing after registration, and it focus some common logic. Use `sortMiddlewares` to arrange the execution order of middleware, likes onion ring model of `koa`.

```javascript
const Koa = require("koa");
// If you use `require`
const { default: HappyRouter } = require("koa-happy-router");
const { default: AsyncValidator } = require("async-validator");

const app = new Koa();
const router = new HappyRouter();

router.registerMiddlewares({
  validate: (config) => async (ctx, next) => {
    const source = { ...ctx.query };
    const schema = new AsyncValidator({ ...config });
    try {
      await schema.validate(source, { first: true });
      await next();
    } catch (error) {
      const { errors = [] } = error;
      ctx.body = { status: -2, ...errors[0] };
    }
  },
  needLogin: (needLogin) => async (ctx, next) => {
    if (needLogin && ctx.state.hasLogin) {
      await next();
    }
    ctx.body = { status: -1, message: "please login" };
  },
});
router.sortMiddlewares(["needLogin", "validate"]);

router.addRoutes([
  {
    url: "/",
    handler: (ctx) => (ctx.body = "Hello World!"),
  },
  {
    url: "/user",
    needLogin: true,
    validate: {
      id: {
        required: true,
      },
    },
    middlewares: [
      async (ctx, next) => {
        console.log("Find user data by id after needLogin and validate");
        await next();
      },
    ],
    handler: (ctx) => (ctx.body = "Hello World!"),
  },
]);

app.use(router.routes()).use(router.allowMethods());
app.listen(8080);
```

### Typescript

`koa-happy-router` is written in Typescript, so you could use it in Typescript.

```typescript
import Koa from "koa";
import HappyRouter from "koa-happy-router";

const app = new Koa();
const router = new HappyRouter();

interface addRoutesProps {
  needLogin: boolean;
}

router.registerMiddlewares({
  needLogin: (needLogin: boolean) => async (ctx, next) => {
    if (needLogin && ctx.state.hasLogin) {
      await next();
    }
    ctx.body = { status: -1, message: "please login" };
  },
});
router.addRoutes<addRoutesProps>([
  {
    url: "/",
    needLogin: true,
    handler: (ctx) => (ctx.body = "Hello World!"),
  },
]);

app.use(router.routes()).use(router.allowMethods());
app.listen(8080);
```


CHANGELOG.md
# 1.1.0 (2021-10-15)

### Bug Fixes

- middlewares is optional field ([b4a94a3](https://github.com/fei-hui/koa-happy-router/commit/b4a94a3eb56098c710c0887712cf63b64e4d4b74))
- update typescript in errorHandler ([d7b9a63](https://github.com/fei-hui/koa-happy-router/commit/d7b9a630e2a4c19e8088934dcc0b795f7e6e1d5b))

### Features

- add log printing ([f76fe7d](https://github.com/fei-hui/koa-happy-router/commit/f76fe7d17dae58481f32366a7e31938db49712f9))
- add new property in constructor ([dd95d0f](https://github.com/fei-hui/koa-happy-router/commit/dd95d0fc65c30a3b8021df915cfabc646214e3ca))

## 1.0.1 (2021-08-19)

### Features

- add new property in constructor ([679d60d](https://github.com/fei-hui/koa-happy-router/commit/679d60d226990067be6af4a4016026f09829ad77))

## 1.0.0 (2021-07-17)

### Features

- chore: the first commit([183a977](https://github.com/fei-hui/koa-happy-router/commit/183a9775aa948b72adfc5eb0d68c8ba53bde391a))


test/index.js
const should = require("should");

describe("Module", () => {
  it("1. Should expose HappyRouter", (done) => {
    const HappyRouter = require("../dist/lib").default;
    should.exist(HappyRouter);
    HappyRouter.should.be.type("function");
    done();
  });
});


test/lib/index.js
const Koa = require("koa");
const http = require("http");
const should = require("should");
const expect = require("expect.js");
const request = require("supertest");
const HappyRouter = require("../../dist/lib").default;

describe("Constructor", () => {
  it("1. Create new HappyRouter", (done) => {
    const router = new HappyRouter();
    router.should.be.instanceOf(HappyRouter);
    done();
  });
  it("2. Add prefix in constructor", (done) => {
    const app = new Koa();
    const router = new HappyRouter({
      prefix: "/api",
    });
    router.addRoutes([
      {
        url: "/",
        method: "GET",
        handler: (ctx) => {
          ctx.body = { text: "Hello World" };
        },
      },
    ]);
    app.use(router.routes()).use(router.allowedMethods());

    request(http.createServer(app.callback()))
      .get("/api")
      .expect(200)
      .end((err, res) => {
        if (err) return done(err);
        expect(res.body).to.eql({ text: "Hello World" });
        done();
      });
  });
  it("3. Add errorHandler in constructor", (done) => {
    const app = new Koa();
    const router = new HappyRouter({
      errorHandler: (error, ctx) => {
        ctx.body = { status: -1, message: error.message };
      },
    });
    router.addRoutes([
      {
        url: "/",
        method: "GET",
        handler: (ctx) => {
          ctx.body = JSON.parse("");
        },
      },
    ]);
    app.use(router.routes()).use(router.allowedMethods());

    request(http.createServer(app.callback()))
      .get("/")
      .expect(200)
      .end((err, res) => {
        if (err) return done(err);
        expect(res.body).to.eql({
          status: -1,
          message: "Unexpected end of JSON input",
        });
        done();
      });
  });
  it("4. Multiple constructor", (done) => {
    const app = new Koa();
    const oneRouter = new HappyRouter({
      prefix: "/one",
    });
    const twoRouter = new HappyRouter({
      prefix: "/two",
    });
    oneRouter.addRoutes([
      {
        url: "/",
        method: "GET",
        handler: (ctx) => (ctx.body = "This is one page"),
      },
    ]);
    twoRouter.addRoutes([
      {
        url: "/",
        method: "GET",
        handler: (ctx) => (ctx.body = "This is two page"),
      },
    ]);
    app.use(oneRouter.routes()).use(twoRouter.routes());

    request(http.createServer(app.callback()))
      .get("/one")
      .expect(200)
      .end((error, response) => {
        if (error) return done(error);
        expect(response.text).to.be("This is one page");

        request(http.createServer(app.callback()))
          .get("/two")
          .expect(200)
          .end((err, res) => {
            if (err) return done(err);
            expect(res.text).to.be("This is two page");
            done();
          });
      });
  });
  it("5. Merge multiple constructor", (done) => {
    const app = new Koa();
    const router = new HappyRouter();
    const oneRouter = new HappyRouter({
      prefix: "/one",
    });
    const twoRouter = new HappyRouter({
      prefix: "/two",
    });
    oneRouter.addRoutes([
      {
        url: "/",
        method: "GET",
        handler: (ctx) => (ctx.body = "This is one page"),
      },
    ]);
    twoRouter.addRoutes([
      {
        url: "/",
        method: "GET",
        handler: (ctx) => (ctx.body = "This is two page"),
      },
    ]);
    router.use(oneRouter.routes()).use(twoRouter.routes());
    app.use(router.routes());

    request(http.createServer(app.callback()))
      .get("/one")
      .expect(200)
      .end((error, response) => {
        if (error) return done(error);
        expect(response.text).to.be("This is one page");

        request(http.createServer(app.callback()))
          .get("/two")
          .expect(200)
          .end((err, res) => {
            if (err) return done(err);
            expect(res.text).to.be("This is two page");
            done();
          });
      });
  });
});

describe("Request", () => {
  it("1. Return text", (done) => {
    const app = new Koa();
    const router = new HappyRouter();
    router.addRoutes([
      {
        url: "/",
        method: "GET",
        handler: (ctx) => {
          ctx.body = "Hello World";
        },
      },
    ]);
    app.use(router.routes()).use(router.allowedMethods());

    request(http.createServer(app.callback()))
      .get("/")
      .expect(200)
      .end((err, res) => {
        if (err) return done(err);
        expect(res.text).to.be("Hello World");
        done();
      });
  });
  it("2. Return HTML", (done) => {
    const app = new Koa();
    const router = new HappyRouter();
    const content = [
      "<!DOCTYPE html>",
      "<head>",
      '  <meta charset="utf-8" />',
      '  <meta content="IE=edge" http-equiv="X-UA-Compatible" />',
      `  <title>Hello, World!</title>`,
      "</head>",
      "<html>",
      "  <body>",
      `    <h1>Hello, World!</h1>`,
      "  </body>",
      "</html>",
    ].join("\t");
    router.addRoutes([
      {
        url: "/",
        method: "GET",
        handler: (ctx) => {
          ctx.body = content;
        },
      },
    ]);
    app.use(router.routes()).use(router.allowedMethods());

    request(http.createServer(app.callback()))
      .get("/")
      .expect(200)
      .end((err, res) => {
        if (err) return done(err);
        expect(res.text).to.be(content);
        done();
      });
  });
  it("3. Return json", (done) => {
    const app = new Koa();
    const router = new HappyRouter();
    router.addRoutes([
      {
        url: "/",
        method: "GET",
        handler: (ctx) => {
          ctx.body = { text: "Hello World" };
        },
      },
    ]);
    app.use(router.routes()).use(router.allowedMethods());

    request(http.createServer(app.callback()))
      .get("/")
      .expect(200)
      .end((err, res) => {
        if (err) return done(err);
        expect(res.body).to.eql({ text: "Hello World" });
        done();
      });
  });
  it("4. Return 404 code", (done) => {
    const app = new Koa();
    const router = new HappyRouter();
    router.addRoutes([
      {
        url: "/",
        method: "GET",
        handler: (ctx) => {
          ctx.code = 404;
        },
      },
    ]);
    app.use(router.routes()).use(router.allowedMethods());

    request(http.createServer(app.callback()))
      .get("/")
      .expect(200)
      .end((err, res) => {
        if (err) {
          expect(res.statusCode).to.be(404);
          done();
        }
      });
  });
  it("5. Mutiple requests", (done) => {
    const app = new Koa();
    const router = new HappyRouter();
    router.addRoutes([
      {
        url: "/",
        method: "GET",
        handler: (ctx) => {
          ctx.body = "Hello World";
        },
      },
      {
        url: "/:name",
        method: "GET",
        handler: (ctx) => {
          ctx.body = `Hello ${ctx.params.name}`;
        },
      },
    ]);
    app.use(router.routes()).use(router.allowedMethods());

    request(http.createServer(app.callback()))
      .get("/")
      .expect(200)
      .end((error, response) => {
        if (error) return done(error);
        expect(response.text).to.be("Hello World");

        request(http.createServer(app.callback()))
          .get("/Feihui")
          .expect(200)
          .end((err, res) => {
            if (err) return done(err);
            expect(res.text).to.be("Hello Feihui");
            done();
          });
      });
  });
  it("6. POST method with parameters", (done) => {
    const app = new Koa();
    const router = new HappyRouter({
      logger: true
    });
    router.addRoutes([
      {
        url: "/demo",
        method: "POST",
        handler: (ctx) => {
          ctx.body = "Hello World";
        },
      },
    ]);
    app.use(router.routes()).use(router.allowedMethods());

    request(http.createServer(app.callback()))
      .post("/demo")
      .set('user-agent', 'should.js')
      .send({ username: 'fei-hui' })
      .expect(200)
      .end((err, res) => {
        if (err) return done(err);
        expect(res.text).to.be("Hello World");
        done();
      });
  });
  it("7. 500 Bad Request with logger", (done) => {
    const app = new Koa();
    const router = new HappyRouter({
      logger: true
    });
    router.addRoutes([
      {
        url: "/",
        method: "GET",
        handler: (ctx) => {
          ctx.status = 500
          ctx.body = JSON.parse("");
        },
      },
    ]);
    app.use(router.routes()).use(router.allowedMethods());

    request(http.createServer(app.callback()))
      .get("/")
      .expect(500)
      .end((err, res) => {
        if(err) return done(err);
        expect(res.text).to.be("Internal Server Error");
        done();
      });
  })
});

describe("Middlewares", () => {
  it("1. Global Middleware", (done) => {
    const app = new Koa();
    const router = new HappyRouter();
    const middlewareSort = [];
    router.registerMiddlewares({
      testMiddlewares: (text) => async (_ctx, next) => {
        middlewareSort.push(text);
        await next();
      },
    });
    router.addRoutes([
      {
        url: "/",
        method: "GET",
        testMiddlewares: "testMiddlewares",
        handler: (ctx) => {
          middlewareSort.push("Hello World");
          ctx.body = middlewareSort.join(",");
        },
      },
    ]);
    app.use(router.routes()).use(router.allowedMethods());

    request(http.createServer(app.callback()))
      .get("/")
      .expect(200)
      .end((err, res) => {
        if (err) return done(err);
        expect(res.text).to.be("testMiddlewares,Hello World");
        done();
      });
  });
  it("2. Multiple global middlewares", (done) => {
    const app = new Koa();
    const router = new HappyRouter();
    const middlewareSort = [];
    router.registerMiddlewares({
      firstMiddlewares: (text) => async (_ctx, next) => {
        middlewareSort.push(text);
        await next();
      },
      secondMiddlewares: (text) => async (_ctx, next) => {
        middlewareSort.push(text);
        await next();
      },
      thirdMiddlewares: (text) => async (_ctx, next) => {
        middlewareSort.push(text);
        await next();
      },
    });
    router.addRoutes([
      {
        url: "/",
        method: "GET",
        firstMiddlewares: "firstMiddlewares",
        secondMiddlewares: "secondMiddlewares",
        thirdMiddlewares: "thirdMiddlewares",
        handler: (ctx) => {
          middlewareSort.push("Hello World");
          ctx.body = middlewareSort.join(",");
        },
      },
    ]);
    app.use(router.routes()).use(router.allowedMethods());

    request(http.createServer(app.callback()))
      .get("/")
      .expect(200)
      .end((err, res) => {
        if (err) return done(err);
        expect(res.text).to.be(
          "firstMiddlewares,secondMiddlewares,thirdMiddlewares,Hello World"
        );
        done();
      });
  });
  it("3. Order of global middlewares", (done) => {
    const app = new Koa();
    const router = new HappyRouter();
    const middlewareSort = [];
    router.registerMiddlewares({
      firstMiddlewares: (text) => async (_ctx, next) => {
        middlewareSort.push(text);
        await next();
      },
      secondMiddlewares: (text) => async (_ctx, next) => {
        middlewareSort.push(text);
        await next();
      },
      thirdMiddlewares: (text) => async (_ctx, next) => {
        middlewareSort.push(text);
        await next();
      },
    });
    router.sortMiddlewares([
      "thirdMiddlewares",
      "firstMiddlewares",
      "secondMiddlewares",
    ]);
    router.addRoutes([
      {
        url: "/",
        method: "GET",
        firstMiddlewares: "firstMiddlewares",
        secondMiddlewares: "secondMiddlewares",
        thirdMiddlewares: "thirdMiddlewares",
        handler: (ctx) => {
          middlewareSort.push("Hello World");
          ctx.body = middlewareSort.join(",");
        },
      },
    ]);
    app.use(router.routes()).use(router.allowedMethods());

    request(http.createServer(app.callback()))
      .get("/")
      .expect(200)
      .end((err, res) => {
        if (err) return done(err);
        expect(res.text).to.be(
          "thirdMiddlewares,firstMiddlewares,secondMiddlewares,Hello World"
        );
        done();
      });
  });
  it("4. Route middlewares", (done) => {
    const app = new Koa();
    const router = new HappyRouter();
    const middlewareSort = [];
    router.addRoutes([
      {
        url: "/",
        method: "GET",
        middlewares: [
          async (_ctx, next) => {
            middlewareSort.push("firstMiddlewaresInRoute");
            await next();
          },
          async (_ctx, next) => {
            middlewareSort.push("secondMiddlewaresInRoute");
            await next();
          },
        ],
        handler: (ctx) => {
          middlewareSort.push("Hello World");
          ctx.body = middlewareSort.join(",");
        },
      },
    ]);
    app.use(router.routes()).use(router.allowedMethods());

    request(http.createServer(app.callback()))
      .get("/")
      .expect(200)
      .end((err, res) => {
        if (err) return done(err);
        expect(res.text).to.be(
          "firstMiddlewaresInRoute,secondMiddlewaresInRoute,Hello World"
        );
        done();
      });
  });
  it("5. Order of middlewares in global and route", (done) => {
    const app = new Koa();
    const router = new HappyRouter();
    const middlewareSort = [];
    router.registerMiddlewares({
      firstMiddlewares: (text) => async (_ctx, next) => {
        middlewareSort.push(text);
        await next();
      },
      secondMiddlewares: (text) => async (_ctx, next) => {
        middlewareSort.push(text);
        await next();
      },
      thirdMiddlewares: (text) => async (_ctx, next) => {
        middlewareSort.push(text);
        await next();
      },
    });
    router.sortMiddlewares([
      "thirdMiddlewares",
      "firstMiddlewares",
      "secondMiddlewares",
    ]);
    router.addRoutes([
      {
        url: "/",
        method: "GET",
        firstMiddlewares: "firstMiddlewares",
        secondMiddlewares: "secondMiddlewares",
        thirdMiddlewares: "thirdMiddlewares",
        middlewares: [
          async (_ctx, next) => {
            middlewareSort.push("firstMiddlewaresInRoute");
            await next();
          },
          async (_ctx, next) => {
            middlewareSort.push("secondMiddlewaresInRoute");
            await next();
          },
        ],
        handler: (ctx) => {
          middlewareSort.push("Hello World");
          ctx.body = middlewareSort.join(",");
        },
      },
    ]);
    app.use(router.routes()).use(router.allowedMethods());

    request(http.createServer(app.callback()))
      .get("/")
      .expect(200)
      .end((err, res) => {
        if (err) return done(err);
        expect(res.text).to.be(
          "thirdMiddlewares,firstMiddlewares,secondMiddlewares,firstMiddlewaresInRoute,secondMiddlewaresInRoute,Hello World"
        );
        done();
      });
  });
  it("6. Only load used middlewares", (done) => {
    const app = new Koa();
    const router = new HappyRouter();
    const middlewareSort = [];
    router.registerMiddlewares({
      firstMiddlewares: (text) => async (_ctx, next) => {
        middlewareSort.push(text);
        await next();
      },
      secondMiddlewares: (text) => async (_ctx, next) => {
        middlewareSort.push(text);
        await next();
      },
      thirdMiddlewares: (text) => async (_ctx, next) => {
        middlewareSort.push(text);
        await next();
      },
    });
    router.addRoutes([
      {
        url: "/",
        method: "GET",
        firstMiddlewares: "firstMiddlewares",
        thirdMiddlewares: "thirdMiddlewares",
        handler: (ctx) => {
          middlewareSort.push("Hello World");
          ctx.body = middlewareSort.join(",");
        },
      },
    ]);
    app.use(router.routes()).use(router.allowedMethods());

    request(http.createServer(app.callback()))
      .get("/")
      .expect(200)
      .end((err, res) => {
        if (err) return done(err);
        expect(res.text).to.be("firstMiddlewares,thirdMiddlewares,Hello World");
        done();
      });
  });
  it("7. Register middleware by use", (done) => {
    const app = new Koa();
    const router = new HappyRouter();
    router.use(async (ctx, next) => {
      ctx.body = ctx.body || { text: "Hello World from use" };
      await next();
    });
    router.addRoutes([
      {
        url: "/helloworld",
        method: "GET",
        handler: (ctx) => {
          ctx.body = { text: "Hello World" };
        },
      },
      {
        url: "/use",
        method: "GET",
      },
    ]);
    app.use(router.routes());
    request(http.createServer(app.callback()))
      .get("/helloworld")
      .expect(200)
      .end((error, response) => {
        if (error) return done(error);
        expect(response.body).to.eql({ text: "Hello World" });

        request(http.createServer(app.callback()))
          .get("/use")
          .expect(200)
          .end((err, res) => {
            if (err) return done(err);
            expect(res.body).to.eql({ text: "Hello World from use" });
            done();
          });
      });
  });
  it("8. Order of middlewares between use and registerMiddlewares", (done) => {
    const app = new Koa();
    const router = new HappyRouter();
    const middlewareSort = [];
    router.use(async (_ctx, next) => {
      middlewareSort.push("beforeOneMiddlewares");
      await next();
    });
    router.registerMiddlewares({
      firstMiddlewares: (text) => async (_ctx, next) => {
        middlewareSort.push(text);
        await next();
      },
      secondMiddlewares: (text) => async (_ctx, next) => {
        middlewareSort.push(text);
        await next();
      },
      thirdMiddlewares: (text) => async (_ctx, next) => {
        middlewareSort.push(text);
        await next();
      },
    });
    router.sortMiddlewares([
      "thirdMiddlewares",
      "firstMiddlewares",
      "secondMiddlewares",
    ]);
    router.use(async (_ctx, next) => {
      middlewareSort.push("beforeTwoMiddlewares");
      await next();
    });
    router.addRoutes([
      {
        url: "/",
        method: "GET",
        firstMiddlewares: "firstMiddlewares",
        secondMiddlewares: "secondMiddlewares",
        thirdMiddlewares: "thirdMiddlewares",
        middlewares: [
          async (_ctx, next) => {
            middlewareSort.push("firstMiddlewaresInRoute");
            await next();
          },
          async (_ctx, next) => {
            middlewareSort.push("secondMiddlewaresInRoute");
            await next();
          },
        ],
        handler: (ctx) => {
          middlewareSort.push("Hello World");
          ctx.body = middlewareSort.join(",");
        },
      },
    ]);
    router.use(async (_ctx, next) => {
      middlewareSort.push("noExecute");
      await next();
    });
    app.use(router.routes()).use(router.allowedMethods());

    request(http.createServer(app.callback()))
      .get("/")
      .expect(200)
      .end((err, res) => {
        if (err) return done(err);
        expect(res.text).to.be(
          "beforeOneMiddlewares,beforeTwoMiddlewares,thirdMiddlewares,firstMiddlewares,secondMiddlewares,firstMiddlewaresInRoute,secondMiddlewaresInRoute,Hello World"
        );
        done();
      });
  });
});


scripts/build.js
const fs = require("fs");
const path = require("path");
const util = require("util");

// 基于项目根目录获取文件
const resolve = (opt) => path.resolve(__dirname, "..", opt);

/**
 * 复制指定的文件到打包文件夹
 *
 * @param {Array<string>} files 需要复制的文件名，文件路径基于项目根目录
 */
const copyFilesToDist = (files) => {
  const copyFile = util.promisify(fs.copyFile);
  return Promise.all(
    files.map((file) => copyFile(resolve(file), resolve(`dist/${file}`)))
  );
};

/**
 * 删除指定的文件
 *
 * @param {Array<string>} files 需要删除的文件，文件路径基于项目根目录
 */
const deleteFiles = (files) => {
  const unlinkFile = util.promisify(fs.unlink);
  return Promise.all(files.map((file) => unlinkFile(resolve(file))));
};

/**
 * 分离开发环境和生产环境的 package.json
 */
const createDistPackage = () => {
  const package = require("../package.json");
  // 修改 package.json 的内容
  {
    package.main = "lib/index.js";
    package.types = "lib/index.d.ts";
    // 删除无用的代码
    delete package.husky;
    delete package.scripts;

    // 移除开发配置中的其他非类型声明的依赖
    if (package.devDependencies) {
      Object.keys(package.devDependencies).forEach((key) => {
        if (!/@types\/(\w)/g.test(key)) {
          delete package.devDependencies[key];
        }
      });
    }
  }
  // 生成新的 package.json
  {
    const filePath = resolve("dist/package.json");
    const fileData = JSON.stringify(package, "", "\t");

    fs.writeFile(filePath, fileData, "utf-8", (err) => {
      if (err) {
        throw new Error(`[Error]`, err.message);
      }
    });
  }
};

createDistPackage();
deleteFiles(["dist/tsconfig.tsbuildinfo"]);
copyFilesToDist([".npmrc", "LICENSE", "README.md"]);


lib/index.ts
import Application from "koa";
import KoaRouter from "koa-router";
import {
  INITIAL_METHOD,
  INITIAL_ROUTE_FIELDS,
  REQUEST_METHODS,
} from "./constants";
import requestLogger from "./logger";

/** Router instance */
export interface RouterOptions {
  /** Prefix for all routes */
  prefix?: string;
  /** Runtime envrionment, it's useless now */
  env?: "development" | "production";
  /** Whether or not routes should matched strictly.
   *  If strict matching is enabled,
   *  the trailing slash is taken into account when matching routes.
   */
  strict?: boolean;
  /**
   * Enable log printing, default is false
   *
   * @example
   * ```shell
   * DEBUG [::ffff:127.0.0.1] [2020-01-01 00:00:00::001] --> GET /api/demo?key=value - UserAgent
   * DEBUG [::ffff:127.0.0.1] [2020-01-01 00:00:00::001] <-- GET /api/demo?key=value - 20ms
   * DEBUG [::ffff:127.0.0.1] [2020-01-01 00:00:00::001] --> POST /api/demo - {key="value"} - UserAgent
   * ERROR [::ffff:127.0.0.1] [2020-01-01 00:00:00::001] <-- POST /api/demo - 500 - 6ms
   * ```
   */
  logger?: boolean;
  /** Methods which should be supported by the router. */
  methods?: Array<keyof typeof REQUEST_METHODS>;
  /** Error collection mechanism to collect error in `handler` method. */
  errorHandler?: ErrorHandler;
}

/** Error collection mechanism to collect error in `handler` method. */
export type ErrorHandler = (
  error: unknown,
  ctx: KoaRouter.RouterContext,
  next?: Application.Next
) => void;

/** Middlewares in route */
export type RouteMiddlewares = Array<
  (ctx: KoaRouter.RouterContext, next: Application.Next) => Promise<void>
>;

/** Handler method in route */
export type RouteHandler = (
  ctx: KoaRouter.RouterContext,
  next?: Application.Next
) => void | Promise<void>;

/** Route properties */
export type Route<T = void> = {
  url: string;
  handler?: RouteHandler;
  middlewares?: RouteMiddlewares;
  method?: keyof typeof REQUEST_METHODS;
} & T;

/** Return middlewares by currying */
export type Middlewares = (
  options: any
) => (ctx: KoaRouter.RouterContext, next: Application.Next) => Promise<void>;

// TODO:
// 1. 测试环境可以自动生成接口文档，可自定义接口文档的地址/页面
/**
 * `koa-happy-router` provides easy-to-use routing configuraion, smart and powerful middleware mechanism.
 * Make the code easy to read and maintain.
 * @property
 * - `addRoutes` - Add routes to router
 * - `sortMiddlewares` - Declare the execution order of middlewares
 * - `registerMiddlewares` - Register middlewares with router
 * - `use` - Use given middleware
 * - `routes` - Returns routes matching the request
 * - `allowedMethods` - Returns separate middleware for response
 * @example
 * ```javascript
 * import Koa from 'koa';
 * import HappyRouter from 'koa-happy-router';
 * const app = new Koa();
 * const router = new HappyRouter();
 *
 * // middlewares
 * router.sortMiddlewares(['auth', 'demo']);
 * router.registerMiddlewares({
 *   demo: (demoOptions = {}) => async (ctx, next) => {
 *     console.log('demo', demoOptions);
 *     await next();
 *   },
 *   auth: (authOptions = false) => async (ctx, next) => {
 *     console.log('auth', authOptions);
 *     await next();
 *   }
 * });
 *
 * // router
 * router.addRoutes([
 *  {
 *    url: '/',
 *    mehtod: 'GET',
 *    auth: true,
 *    demo: {
 *      text: 'auth will excute before than demo'
 *    },
 *    handler: (ctx) => ctx.body = `Hello World`
 *  }
 * ]);
 * app.use(router.routes()).use(router.allowedMethods());
 *
 * app.listen(8080);
 * ```
 */
class HappyRouter {
  private enableLogger: boolean;
  private errorHandler?: ErrorHandler;
  private routerInstance?: KoaRouter;
  private middlewaresKeys: Set<string> = new Set();
  private middlewaresStack: Map<string, Middlewares> = new Map();

  /**
   * @property
   * - `env` Runtime envrionment.
   * - `prefix` Prefix for all routes.
   * - `logger` Enable log printing.
   * - `strict` Whether or not routes should matched strictly. If `strict` matching is enabled, the trailing slash is taken into account when matching routes.
   * - `methods` Methods which should be supported by the router.
   * - `errorHandler` Error collection mechanism to collect error in `handler` method.
   * @example
   * ```javascript
   * import Koa from 'koa';
   * import HappyRouter from 'koa-happy-router';
   * const app = new Koa();
   * const router = new HappyRouter({
   *   prefix: '/api',
   *   env: 'development',
   *   strict: true,
   *   logger: true,
   *   methods: ['GET', 'POST', 'OPTIONS'],
   *   errorHandler: (error, ctx) => {
   *     ctx.body = { status: -1, message: error.message, url: ctx.url }
   *   }
   * });
   * ```
   */
  constructor(options?: RouterOptions) {
    if (typeof options?.errorHandler === "function") {
      this.errorHandler = options.errorHandler;
    }
    this.routerInstance = new KoaRouter({
      prefix: options?.prefix,
      strict: options?.strict,
      methods: options?.methods,
    });
    // Add log print
    this.enableLogger = !!options?.logger;
    if (this.enableLogger) {
      this.routerInstance.use(requestLogger);
    }
  }

  /**
   * Add optional error collection mechanism to collect error in `handler` method.
   */
  private async createHandler(
    handler: RouteHandler,
    context: KoaRouter.RouterContext,
    next: Application.Next
  ): Promise<void> {
    if (this.errorHandler) {
      try {
        await handler(context, next);
      } catch (error) {
        this.errorHandler(error, context);
      }
    } else {
      await handler(context, next);
    }
  }
  /**
   * Add routes to router, you could handle the logic in middlewares
   * if middlewares registered by `registerMiddlewares` method.
   *
   * @example
   * ```javascript
   * router.addRoutes([
   *   // Basic Example
   *   {
   *     url: '/',
   *     method: 'GET',
   *     handler: ctx => ctx.body = `Hello, World!`
   *   },
   *   // Middlewares Example
   *   {
   *     url: '/user',
   *     method: 'POST',
   *     permit: true, // Middlewares in global register by registerMiddlewares
   *     middlewares: [
   *       // Middlewares in route
   *       async (ctx, next) => {
   *         console.log('Route middlewares');
   *         await next();
   *       }
   *     ],
   *     handler: ctx => ctx.body = `Hello, ${ctx.session.name}!`
   *   }
   * ]);
   * ```
   */
  public addRoutes<T>(routesList: Route<T>[]): void {
    const routerInstance = this.routerInstance as KoaRouter;
    routesList.forEach((route) => {
      const routeMethod = REQUEST_METHODS[route.method || INITIAL_METHOD];
      const routeInstance = routerInstance[routeMethod].bind(routerInstance);

      // Filter middlewares which unregister in global and unused in route
      const routeMiddlewaresKeys = Object.keys(route).filter(
        (key) =>
          !INITIAL_ROUTE_FIELDS.includes(key) && this.middlewaresStack.has(key)
      );
      // The order of declaration in sortMiddlewares prefer than keys from Object.keys when isn't declared.
      const middlewaresKeys =
        this.middlewaresKeys.size > 0
          ? [...this.middlewaresKeys].filter((key) =>
              routeMiddlewaresKeys.includes(key)
            )
          : routeMiddlewaresKeys;

      // Global middleware takes precedence over local middleware
      const routeMiddlewares = [
        ...middlewaresKeys.map((key) => {
          const middlewareInstance = this.middlewaresStack.get(
            key
          ) as Middlewares;
          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
          // @ts-ignore
          return middlewareInstance(route[key]);
        }),
        ...(route.middlewares || []),
      ];

      routeInstance(
        route.url,
        ...(routeMiddlewares as KoaRouter.IMiddleware[]),
        (ctx, next) =>
          route.handler && this.createHandler(route.handler, ctx, next)
      );
    });
  }
  /**
   * Declare the execution order of middlewares and execute in this order.
   * Used earlier than the `addRoutes` method.
   *
   * @example
   * ```javascript
   * router.sortMiddlewares(['permit', 'validate']);
   * ```
   */
  public sortMiddlewares(keys: string[] = []): void {
    keys.forEach((key) => {
      this.middlewaresKeys.add(key);
    });
  }
  /**
   * Register middlewares with router, and its will load only used middlewares in routes.
   * Used earlier than the `addRoutes` method.
   *
   * @example
   * ```javascript
   * router.registerMiddlewares({
   *   permit: (options = false) => async (ctx, next) => {
   *     if(options && !!ctx.session.status) {
   *       return ctx.body = 'please login';
   *     }
   *     await next();
   *   }
   * });
   * ```
   */
  public registerMiddlewares(middlewares: Record<string, Middlewares>): void {
    Object.keys(middlewares).forEach((key) => {
      this.middlewaresStack.set(key, middlewares[key]);
    });
  }
  /**
   *
   * Use constuctor or middlewares.
   *
   * Middlewares will execute in the order of defined by `.use()` and its run before `registerMiddlewares`.
   * Used earlier than the `addRoutes` method.
   *
   * @example
   * ```javascript
   * router.use(async(ctx, next) => {
   *   console.log('use before registerMiddlewares');
   *   await next();
   * });
   * ```
   */
  public use(...middleware: RouteMiddlewares) {
    const routerInstance = this.routerInstance as KoaRouter;
    return routerInstance.use(...middleware);
  }
  /**
   * Returns router middleware which dispatches a route matching the request.
   */
  public routes(): KoaRouter.IMiddleware {
    const routerInstance = this.routerInstance as KoaRouter;
    return routerInstance.routes();
  }
  /**
   * Returns separate middleware for responding to `OPTIONS` requests with an `Allow` header containing the allowed methods,
   * as well as responding with `405 Method Not Allowed` and `501 Not Implemented` as appropriate.
   */
  public allowedMethods(
    options?: KoaRouter.IRouterAllowedMethodsOptions
  ): KoaRouter.IMiddleware {
    const routerInstance = this.routerInstance as KoaRouter;
    return routerInstance.allowedMethods(options);
  }
}

export default HappyRouter;
